File: src/bdf/distributions/Poisson.py
"""Poisson distribution implementations for BDF.

Provides conjugate Poisson-Gamma models:
- GammaABLambdaPoisson: Gamma(α, β) prior on rate λ (shape-rate parameterization)
- GammaMVLambdaPoisson: Gamma prior via mean/variance (more interpretable)

Both support:
- Closed-form Bayesian evidence (NLE)
- Negative Binomial posterior predictive
- Efficient inference (conjugate updates)
"""
from typing import ClassVar, Literal

import numpy as np
from pydantic import Field
from scipy.stats import gamma as gamma_dist
from scipy.stats import nbinom, poisson

from bdf.distributions.bdf_distribution import BDFDistribution, BDFDistributionParams
from bdf.utils.constants import RANDOM_SEED

# ============================================================================
# PARAMS CLASSES
# ============================================================================


class GammaABLambdaPoissonParams(BDFDistributionParams):
    """Parameters for Poisson with Gamma(α, β) prior on rate λ.

    Prior: λ ~ Gamma(α, β)  [shape-rate parameterization]
    Likelihood: y | λ ~ Poisson(λ)
    Posterior: λ | y ~ Gamma(α + Σy, β + n)
    Posterior predictive: y_new | y ~ NegativeBinomial(α_post, β_post/(1+β_post))

    Notes:
    - α (alpha_lambda): shape parameter (α > 0)
    - β (beta_lambda): rate parameter (β > 0)
    - E[λ] = α/β, Var[λ] = α/β²
    - NB parameterization: n=α_post, p=β_post/(1+β_post)
    """

    # Prior hyperparameters
    alpha_lambda: float = Field(default=1.0, gt=0, description="Shape parameter α of Gamma prior for rate λ")
    beta_lambda: float = Field(default=1.0, gt=0, description="Rate parameter β of Gamma prior for rate λ")

    # Scoring defaults for conjugate model
    score_method: Literal["nle", "nll"] = Field(
        default="nle", description="Conjugate model defaults to nle (Bayesian evidence)."
    )
    use_posterior_predictive: bool = Field(
        default=True, description="Use Negative Binomial posterior predictive (marginalizes λ uncertainty)."
    )


class GammaMVLambdaPoissonParams(BDFDistributionParams):
    """Parameters for Poisson with Gamma prior specified via mean/variance.

    Prior: λ ~ Gamma(α, β) where α = mean²/var, β = mean/var
    Likelihood: y | λ ~ Poisson(λ)
    Posterior: λ | y ~ Gamma(α + Σy, β + n)
    Posterior predictive: y_new | y ~ NegativeBinomial(α_post, β_post/(1+β_post))

    This parameterization is more interpretable:
    - mean_lambda: prior expectation of rate λ
    - var_lambda: prior uncertainty about λ
    """

    # Prior hyperparameters (mean-variance parameterization)
    mean_lambda: float = Field(default=1.0, gt=0, description="Prior mean E[λ] for rate parameter")
    var_lambda: float = Field(default=1.0, gt=0, description="Prior variance Var[λ] for rate parameter")

    # Scoring defaults
    score_method: Literal["nle", "nll"] = Field(default="nle")
    use_posterior_predictive: bool = Field(default=True, description="Use Negative Binomial posterior predictive.")


# ============================================================================
# DISTRIBUTION IMPLEMENTATIONS
# ============================================================================


class GammaABLambdaPoisson(BDFDistribution):
    """Poisson-Gamma conjugate model with shape-rate (α, β) parameterization.

    Supports:
    - Closed-form Bayesian evidence (NLE)
    - Negative Binomial posterior predictive (integrates out λ uncertainty)
    - Efficient conjugate updates
    """

    params_cls: ClassVar[type[BDFDistributionParams]] = GammaABLambdaPoissonParams

    # Capabilities
    _supports_nle = True
    _has_fast_loo_cv = False
    _has_fast_kfold_cv = False
    _supports_posterior_predictive = True

    def __init__(self, params: GammaABLambdaPoissonParams):
        super().__init__(params)
        self.alpha_lambda = params.alpha_lambda
        self.beta_lambda = params.beta_lambda

    # ========================================================================
    # REQUIRED METHODS
    # ========================================================================

    def calc_posterior_params(self, data: np.ndarray) -> dict[str, float]:
        """Calculate Gamma posterior parameters for λ.

        Returns dict with:
        - posterior_alpha: Posterior shape α_post = α + Σy
        - posterior_beta: Posterior rate β_post = β + n
        - posterior_lambda: Posterior mean E[λ | y] = α_post/β_post
        """
        n = data.shape[0]
        sum_counts = np.sum(data)

        # Conjugate update: Gamma(α, β) → Gamma(α + Σy, β + n)
        alpha_post = self.alpha_lambda + sum_counts
        beta_post = self.beta_lambda + n

        # Posterior mean of λ
        lambda_post = alpha_post / beta_post

        return {
            "posterior_alpha": float(alpha_post),
            "posterior_beta": float(beta_post),
            "posterior_lambda": float(lambda_post),
        }

    def _plugin_log_likelihood(self, data: np.ndarray, params: dict) -> np.ndarray:
        """Plug-in Poisson likelihood: Poisson(k | λ_post).

        Uses posterior mean λ = α_post/β_post as point estimate.
        """
        lambda_post = params["posterior_lambda"]
        # Poisson PMF: λ^k exp(-λ) / k!
        return poisson.logpmf(data, mu=lambda_post)

    def _num_parameters(self) -> int:
        """Only λ is estimated."""
        return 1

    def _sample_posterior_params(self, params: dict[str, float], size: int, random_state: int) -> np.ndarray:
        """Sample from posterior predictive (Negative Binomial distribution).

        If use_posterior_predictive=True:
            Sample from NegativeBinomial(n=α_post, p=β_post/(1+β_post))
        Else:
            Sample from Poisson(λ_post)
        """
        if self.params.use_posterior_predictive:
            # Posterior predictive: Negative Binomial
            alpha_post = params["posterior_alpha"]
            beta_post = params["posterior_beta"]

            # scipy.stats.nbinom: n (number of successes), p (success probability)
            # Conversion: n = α_post, p = β_post/(1+β_post)
            n_nb = alpha_post
            p_nb = beta_post / (1 + beta_post)

            return np.array(nbinom.rvs(n=n_nb, p=p_nb, size=size, random_state=random_state))
        else:
            # Plug-in: Poisson(λ_post)
            lambda_post = params["posterior_lambda"]
            return np.array(poisson.rvs(mu=lambda_post, size=size, random_state=random_state))

    def validate_targets(self, data: np.ndarray):
        """Validate data for Poisson distribution."""
        if data.ndim != 1:
            raise ValueError(f"Data must be 1-dimensional, got shape {data.shape}")
        if len(data) == 0:
            raise ValueError("Data cannot be empty")
        if not np.all(data >= 0):
            raise ValueError("Poisson data must be non-negative integers (k ≥ 0)")
        if not np.all(data == np.floor(data)):
            raise ValueError("Poisson data must be integers")
        if not np.all(np.isfinite(data)):
            raise ValueError("Data contains non-finite values")

    def get_posterior_mean(self, *, data: np.ndarray | None = None, params: dict[str, float] | None = None) -> float:
        """Get posterior mean of the distribution.

        For Poisson(λ), the mean is λ.
        For posterior predictive NB(α, β), mean = α(1-p)/p = α/β.
        """
        if params is None:
            if data is None:
                raise ValueError("Provide either 'data' or 'params'")
            params = self.calc_posterior_params(data)

        if self.params.use_posterior_predictive:
            # Negative Binomial mean: n(1-p)/p = α/β
            alpha_post = params["posterior_alpha"]
            beta_post = params["posterior_beta"]
            return alpha_post / beta_post
        else:
            # Poisson mean: λ
            return params["posterior_lambda"]

    def get_posterior_variance(
        self, *, data: np.ndarray | None = None, params: dict[str, float] | None = None
    ) -> float:
        """Get posterior variance of the distribution.

        For Poisson(λ), variance = λ.
        For posterior predictive NB(α, β), variance = α(1-p)/p² = α(1+β)/β².
        """
        if params is None:
            if data is None:
                raise ValueError("Provide either 'data' or 'params'")
            params = self.calc_posterior_params(data)

        if self.params.use_posterior_predictive:
            # Negative Binomial variance: n(1-p)/p² = α(1+β)/β²
            alpha_post = params["posterior_alpha"]
            beta_post = params["posterior_beta"]
            return alpha_post * (1 + beta_post) / (beta_post**2)
        else:
            # Poisson variance: λ
            return params["posterior_lambda"]

    # ========================================================================
    # OPTIONAL METHODS (OVERRIDE FOR EFFICIENCY)
    # ========================================================================

    def log_evidence(self, data: np.ndarray) -> float:
        """Exact Bayesian evidence for Poisson-Gamma conjugate.

        p(y | prior) = ∫ p(y | λ) p(λ) dλ

        Closed form for Gamma-Poisson:
        log p(y) = log Γ(α + Σy) - log Γ(α) + α log β - (α + Σy) log(β + n) - Σ log(y!)
        """
        from scipy.special import gammaln

        n = data.shape[0]
        sum_counts = np.sum(data)

        alpha_post = self.alpha_lambda + sum_counts
        beta_post = self.beta_lambda + n

        # Log evidence = log marginal likelihood
        log_ev = gammaln(alpha_post) - gammaln(self.alpha_lambda)
        log_ev += self.alpha_lambda * np.log(self.beta_lambda)
        log_ev -= alpha_post * np.log(beta_post)

        # Subtract log factorials (Poisson normalization)
        log_ev -= np.sum(gammaln(data + 1))

        return float(log_ev)

    def _posterior_predictive_log_likelihood(self, data: np.ndarray, params: dict) -> np.ndarray:
        """Posterior predictive: Negative Binomial(α_post, β_post/(1+β_post)).

        Integrates out uncertainty in λ over the Gamma posterior.

        NB PMF: f(k | n, p) = C(k+n-1, k) p^n (1-p)^k
        where n = α_post, p = β_post/(1+β_post)
        """
        alpha_post = params["posterior_alpha"]
        beta_post = params["posterior_beta"]

        # scipy.stats.nbinom: n (number of successes), p (success probability)
        n_nb = alpha_post
        p_nb = beta_post / (1 + beta_post)

        return nbinom.logpmf(data, n=n_nb, p=p_nb)

    def sample_prior(self, size: int, random_state: int = RANDOM_SEED) -> np.ndarray:
        """Sample λ from prior Gamma(α, β).

        Returns rate parameters λ (not Poisson samples).
        To get prior predictive samples, draw λ ~ Gamma(α, β) then k ~ Poisson(λ).
        """
        rng = np.random.default_rng(random_state)
        # scipy.stats.gamma: a=α (shape), scale=1/β (rate→scale conversion)
        return np.array(gamma_dist.rvs(a=self.alpha_lambda, scale=1.0 / self.beta_lambda, size=size, random_state=rng))


class GammaMVLambdaPoisson(BDFDistribution):
    """Poisson-Gamma conjugate model with mean-variance parameterization.

    Same as GammaABLambdaPoisson, but prior specified via:
    - mean_lambda = E[λ]
    - var_lambda = Var[λ]

    Internally converts to α = mean²/var, β = mean/var.
    """

    params_cls: ClassVar[type[BDFDistributionParams]] = GammaMVLambdaPoissonParams

    _supports_nle = True
    _has_fast_loo_cv = False
    _has_fast_kfold_cv = False
    _supports_posterior_predictive = True

    def __init__(self, params: GammaMVLambdaPoissonParams):
        super().__init__(params)
        self.mean_lambda = params.mean_lambda
        self.var_lambda = params.var_lambda

        # Convert mean-variance to shape-rate
        self.alpha_lambda = self.mean_lambda**2 / self.var_lambda
        self.beta_lambda = self.mean_lambda / self.var_lambda

    # ========================================================================
    # REQUIRED METHODS (identical to GammaABLambdaPoisson after conversion)
    # ========================================================================

    def calc_posterior_params(self, data: np.ndarray) -> dict[str, float]:
        """Calculate Gamma posterior parameters."""
        n = data.shape[0]
        sum_counts = np.sum(data)

        alpha_post = self.alpha_lambda + sum_counts
        beta_post = self.beta_lambda + n
        lambda_post = alpha_post / beta_post

        return {
            "posterior_alpha": float(alpha_post),
            "posterior_beta": float(beta_post),
            "posterior_lambda": float(lambda_post),
        }

    def _plugin_log_likelihood(self, data: np.ndarray, params: dict) -> np.ndarray:
        """Plug-in Poisson likelihood."""
        lambda_post = params["posterior_lambda"]
        return poisson.logpmf(data, mu=lambda_post)

    def _num_parameters(self) -> int:
        return 1

    def _sample_posterior_params(self, params: dict[str, float], size: int, random_state: int) -> np.ndarray:
        """Sample from posterior predictive or plug-in."""
        if self.params.use_posterior_predictive:
            alpha_post = params["posterior_alpha"]
            beta_post = params["posterior_beta"]

            n_nb = alpha_post
            p_nb = beta_post / (1 + beta_post)

            return np.array(nbinom.rvs(n=n_nb, p=p_nb, size=size, random_state=random_state))
        else:
            lambda_post = params["posterior_lambda"]
            return np.array(poisson.rvs(mu=lambda_post, size=size, random_state=random_state))

    def validate_targets(self, data: np.ndarray):
        """Validate data."""
        if data.ndim != 1:
            raise ValueError(f"Data must be 1-dimensional, got shape {data.shape}")
        if len(data) == 0:
            raise ValueError("Data cannot be empty")
        if not np.all(data >= 0):
            raise ValueError("Poisson data must be non-negative integers (k ≥ 0)")
        if not np.all(data == np.floor(data)):
            raise ValueError("Poisson data must be integers")
        if not np.all(np.isfinite(data)):
            raise ValueError("Data contains non-finite values")

    def get_posterior_mean(self, *, data: np.ndarray | None = None, params: dict[str, float] | None = None) -> float:
        """Get posterior mean."""
        if params is None:
            if data is None:
                raise ValueError("Provide either 'data' or 'params'")
            params = self.calc_posterior_params(data)

        if self.params.use_posterior_predictive:
            alpha_post = params["posterior_alpha"]
            beta_post = params["posterior_beta"]
            return alpha_post / beta_post
        else:
            return params["posterior_lambda"]

    def get_posterior_variance(
        self, *, data: np.ndarray | None = None, params: dict[str, float] | None = None
    ) -> float:
        """Get posterior variance."""
        if params is None:
            if data is None:
                raise ValueError("Provide either 'data' or 'params'")
            params = self.calc_posterior_params(data)

        if self.params.use_posterior_predictive:
            alpha_post = params["posterior_alpha"]
            beta_post = params["posterior_beta"]
            return alpha_post * (1 + beta_post) / (beta_post**2)
        else:
            return params["posterior_lambda"]

    # ========================================================================
    # OPTIONAL METHODS
    # ========================================================================

    def log_evidence(self, data: np.ndarray) -> float:
        """Exact Bayesian evidence."""
        from scipy.special import gammaln

        n = data.shape[0]
        sum_counts = np.sum(data)

        alpha_post = self.alpha_lambda + sum_counts
        beta_post = self.beta_lambda + n

        log_ev = gammaln(alpha_post) - gammaln(self.alpha_lambda)
        log_ev += self.alpha_lambda * np.log(self.beta_lambda)
        log_ev -= alpha_post * np.log(beta_post)
        log_ev -= np.sum(gammaln(data + 1))

        return float(log_ev)

    def _posterior_predictive_log_likelihood(self, data: np.ndarray, params: dict) -> np.ndarray:
        """Posterior predictive: Negative Binomial."""
        alpha_post = params["posterior_alpha"]
        beta_post = params["posterior_beta"]

        n_nb = alpha_post
        p_nb = beta_post / (1 + beta_post)

        return nbinom.logpmf(data, n=n_nb, p=p_nb)

    def sample_prior(self, size: int, random_state: int = RANDOM_SEED) -> np.ndarray:
        """Sample λ from prior Gamma(α, β)."""
        rng = np.random.default_rng(random_state)
        return np.array(gamma_dist.rvs(a=self.alpha_lambda, scale=1.0 / self.beta_lambda, size=size, random_state=rng))

File: compiled/rust/src/distributions/poisson.rs
use super::{DistributionPrimitives, SufficientStats};
use ndarray::{ArrayView1, Array1};
use pyo3::prelude::*;
use pyo3::types::PyDict;
use std::collections::HashMap;
use std::f64::consts::PI;

// Lanczos approximation for log-gamma function
fn lgamma(x: f64) -> f64 {
    let p = [
        0.99999999999980993, 676.5203681218851, -1259.1392167224028,
        771.32342877765313, -176.61502916214059, 12.507343278686905,
        -0.13857109526572012, 9.9843695780195716e-6, 1.5056327351493116e-7
    ];
    let g = 7.0;
    if x < 0.5 {
        PI.ln() - (PI * x).sin().ln() - lgamma(1.0 - x)
    } else {
        let z = x - 1.0;
        let mut a = p[0];
        for i in 1..9 {
            a += p[i] / (z + i as f64);
        }
        let t = z + g + 0.5;
        0.5 * (2.0 * PI).ln() + (z + 0.5) * t.ln() - t + a.ln()
    }
}

// ============================================================================
// GammaABLambdaPoisson (Shape-Rate Parameterization)
// ============================================================================

/// Gamma-Poisson conjugate model with shape-rate (α, β) parameterization.
///
/// Prior: λ ~ Gamma(alpha_lambda, beta_lambda)  [shape-rate]
/// Likelihood: y | λ ~ Poisson(λ)
/// Posterior: λ | y ~ Gamma(alpha_lambda + Σy, beta_lambda + n)
/// Posterior Predictive: y_new | y ~ NegativeBinomial(α_post, β_post/(1+β_post))
pub struct GammaABLambdaPoisson {
    pub alpha_lambda: f64,  // Shape parameter
    pub beta_lambda: f64,   // Rate parameter
}

impl GammaABLambdaPoisson {
    pub fn from_spec(spec: &PyDict) -> PyResult<Self> {
        Ok(Self {
            alpha_lambda: spec.get_item("alpha_lambda").ok_or_else(|| pyo3::exceptions::PyValueError::new_err(
                "Missing alpha_lambda in GammaABLambdaPoisson spec"
            ))?.extract()?,
            beta_lambda: spec.get_item("beta_lambda").ok_or_else(|| pyo3::exceptions::PyValueError::new_err(
                "Missing beta_lambda in GammaABLambdaPoisson spec"
            ))?.extract()?,
        })
    }
}

impl DistributionPrimitives for GammaABLambdaPoisson {
    fn calc_posterior_params(&self, data: &ArrayView1<f64>) -> HashMap<String, f64> {
        let n = data.len() as f64;
        let sum_y: f64 = data.iter().sum();

        // Conjugate update: Gamma(α, β) → Gamma(α + Σy, β + n)
        let alpha_post = self.alpha_lambda + sum_y;
        let beta_post = self.beta_lambda + n;

        // Posterior mean of λ
        let lambda_post = alpha_post / beta_post;

        let mut params = HashMap::new();
        params.insert("posterior_alpha".to_string(), alpha_post);
        params.insert("posterior_beta".to_string(), beta_post);
        params.insert("posterior_lambda".to_string(), lambda_post);
        params
    }

    fn plugin_log_likelihood(&self, data: &ArrayView1<f64>, params: &HashMap<String, f64>) -> Array1<f64> {
        let lambda = params["posterior_lambda"].max(1e-10);
        let log_lambda = lambda.ln();

        // Poisson PMF: log P(k | λ) = k*log(λ) - λ - log(k!)
        data.mapv(|k| {
            k * log_lambda - lambda - lgamma(k + 1.0)
        })
    }

    fn posterior_predictive_log_likelihood(&self, data: &ArrayView1<f64>, params: &HashMap<String, f64>) -> Option<Array1<f64>> {
        // Posterior predictive is Negative Binomial
        // NB(k | r, p) where r = α_post, p = β_post / (1 + β_post)
        let alpha_post = params["posterior_alpha"];
        let beta_post = params["posterior_beta"];

        let r = alpha_post;
        let p = beta_post / (1.0 + beta_post);
        let log_p = p.ln();
        let log_1mp = (1.0 - p).ln();

        // NB PMF: log P(k | r, p) = log Γ(k+r) - log Γ(k+1) - log Γ(r) + r*log(p) + k*log(1-p)
        Some(data.mapv(|k| {
            lgamma(k + r) - lgamma(k + 1.0) - lgamma(r) + r * log_p + k * log_1mp
        }))
    }

    fn nle(&self, data: &ArrayView1<f64>) -> Option<f64> {
        let n = data.len() as f64;
        if n == 0.0 { return Some(0.0); }

        let sum_y: f64 = data.iter().sum();

        let alpha_post = self.alpha_lambda + sum_y;
        let beta_post = self.beta_lambda + n;

        // Log Evidence for Gamma-Poisson:
        // log p(y) = log Γ(α_post) - log Γ(α) + α*log(β) - α_post*log(β_post) - Σ log(y!)
        let log_evidence = lgamma(alpha_post) - lgamma(self.alpha_lambda)
            + self.alpha_lambda * self.beta_lambda.ln()
            - alpha_post * beta_post.ln()
            - data.iter().map(|&k| lgamma(k + 1.0)).sum::<f64>();

        Some(-log_evidence) // Return NLE
    }

    fn nll(&self, data: &ArrayView1<f64>, use_posterior_predictive: bool) -> f64 {
        let n = data.len() as f64;
        if n == 0.0 { return 0.0; }

        let sum_y: f64 = data.iter().sum();

        let alpha_post = self.alpha_lambda + sum_y;
        let beta_post = self.beta_lambda + n;

        if use_posterior_predictive {
            // Negative Binomial
            let r = alpha_post;
            let p = beta_post / (1.0 + beta_post);
            let log_p = p.ln();
            let log_1mp = (1.0 - p).ln();

            let mut nll = 0.0;
            for &k in data {
                let log_pmf = lgamma(k + r) - lgamma(k + 1.0) - lgamma(r) + r * log_p + k * log_1mp;
                nll -= log_pmf;
            }
            nll
        } else {
            // Plug-in Poisson
            let lambda = alpha_post / beta_post;
            let log_lambda = lambda.max(1e-10).ln();

            let mut nll = 0.0;
            for &k in data {
                let log_pmf = k * log_lambda - lambda - lgamma(k + 1.0);
                nll -= log_pmf;
            }
            nll
        }
    }

    fn nll_train_test(&self, train: &ArrayView1<f64>, test: &ArrayView1<f64>, use_posterior_predictive: bool) -> f64 {
        let n_train = train.len() as f64;
        let sum_y_train: f64 = train.iter().sum();

        let alpha_post = self.alpha_lambda + sum_y_train;
        let beta_post = self.beta_lambda + n_train;

        if use_posterior_predictive {
            let r = alpha_post;
            let p = beta_post / (1.0 + beta_post);
            let log_p = p.ln();
            let log_1mp = (1.0 - p).ln();

            let mut nll = 0.0;
            for &k in test {
                let log_pmf = lgamma(k + r) - lgamma(k + 1.0) - lgamma(r) + r * log_p + k * log_1mp;
                nll -= log_pmf;
            }
            nll
        } else {
            let lambda = alpha_post / beta_post;
            let log_lambda = lambda.max(1e-10).ln();

            let mut nll = 0.0;
            for &k in test {
                let log_pmf = k * log_lambda - lambda - lgamma(k + 1.0);
                nll -= log_pmf;
            }
            nll
        }
    }

    // ========================================================================
    // FAST PATH: Sufficient Statistics
    // ========================================================================

    fn nll_suff_stats(&self, stats: &SufficientStats, use_posterior_predictive: bool) -> Option<f64> {
        let n = stats.n;
        if n == 0.0 { return Some(0.0); }

        // For Poisson: sum = Σy (sum of counts)
        let sum_y = stats.sum;

        let alpha_post = self.alpha_lambda + sum_y;
        let beta_post = self.beta_lambda + n;

        if use_posterior_predictive {
            // Negative Binomial NLL requires individual k values for lgamma(k + r)
            // Cannot be computed from sufficient stats alone
            return None;
        } else {
            // Plug-in Poisson: NLL = n*λ - Σy*log(λ) + Σlog(y!)
            // The Σlog(y!) term requires individual data points, so we cannot compute it exactly.
            // However, for SPLIT COMPARISON, this term is constant and cancels out!
            // We return the comparable part only.
            let lambda = (alpha_post / beta_post).max(1e-10);
            let log_lambda = lambda.ln();

            // NLL (comparable part) = n*λ - Σy*log(λ)
            // Note: We omit -Σlog(y!) because it's constant across splits
            let nll = n * lambda - sum_y * log_lambda;
            Some(nll)
        }
    }

    fn nle_suff_stats(&self, stats: &SufficientStats) -> Option<f64> {
        let n = stats.n;
        if n == 0.0 { return Some(0.0); }

        let sum_y = stats.sum;

        let alpha_post = self.alpha_lambda + sum_y;
        let beta_post = self.beta_lambda + n;

        // Log Evidence (comparable part, omitting -Σlog(y!)):
        // log p(y) = log Γ(α_post) - log Γ(α) + α*log(β) - α_post*log(β_post)
        // The -Σlog(y!) term cancels when comparing splits
        let log_evidence = lgamma(alpha_post) - lgamma(self.alpha_lambda)
            + self.alpha_lambda * self.beta_lambda.ln()
            - alpha_post * beta_post.ln();

        Some(-log_evidence)
    }
}

// ============================================================================
// GammaMVLambdaPoisson (Mean-Variance Parameterization)
// ============================================================================

/// Gamma-Poisson conjugate model with mean-variance parameterization.
///
/// Prior specified via:
/// - mean_lambda = E[λ] (prior expected rate)
/// - var_lambda = Var[λ] (prior uncertainty about rate)
///
/// Internally converts to α = mean²/var, β = mean/var
pub struct GammaMVLambdaPoisson {
    pub mean_lambda: f64,
    pub var_lambda: f64,
    // Derived shape-rate parameters
    alpha_lambda: f64,
    beta_lambda: f64,
}

impl GammaMVLambdaPoisson {
    pub fn from_spec(spec: &PyDict) -> PyResult<Self> {
        let mean_lambda: f64 = spec.get_item("mean_lambda").ok_or_else(|| pyo3::exceptions::PyValueError::new_err(
            "Missing mean_lambda in GammaMVLambdaPoisson spec"
        ))?.extract()?;
        let var_lambda: f64 = spec.get_item("var_lambda").ok_or_else(|| pyo3::exceptions::PyValueError::new_err(
            "Missing var_lambda in GammaMVLambdaPoisson spec"
        ))?.extract()?;

        // Convert mean-variance to shape-rate
        // E[λ] = α/β, Var[λ] = α/β²
        // → α = mean²/var, β = mean/var
        let alpha_lambda = mean_lambda * mean_lambda / var_lambda;
        let beta_lambda = mean_lambda / var_lambda;

        Ok(Self {
            mean_lambda,
            var_lambda,
            alpha_lambda,
            beta_lambda,
        })
    }
}

impl DistributionPrimitives for GammaMVLambdaPoisson {
    fn calc_posterior_params(&self, data: &ArrayView1<f64>) -> HashMap<String, f64> {
        let n = data.len() as f64;
        let sum_y: f64 = data.iter().sum();

        let alpha_post = self.alpha_lambda + sum_y;
        let beta_post = self.beta_lambda + n;
        let lambda_post = alpha_post / beta_post;

        let mut params = HashMap::new();
        params.insert("posterior_alpha".to_string(), alpha_post);
        params.insert("posterior_beta".to_string(), beta_post);
        params.insert("posterior_lambda".to_string(), lambda_post);
        params
    }

    fn plugin_log_likelihood(&self, data: &ArrayView1<f64>, params: &HashMap<String, f64>) -> Array1<f64> {
        let lambda = params["posterior_lambda"].max(1e-10);
        let log_lambda = lambda.ln();

        data.mapv(|k| {
            k * log_lambda - lambda - lgamma(k + 1.0)
        })
    }

    fn posterior_predictive_log_likelihood(&self, data: &ArrayView1<f64>, params: &HashMap<String, f64>) -> Option<Array1<f64>> {
        let alpha_post = params["posterior_alpha"];
        let beta_post = params["posterior_beta"];

        let r = alpha_post;
        let p = beta_post / (1.0 + beta_post);
        let log_p = p.ln();
        let log_1mp = (1.0 - p).ln();

        Some(data.mapv(|k| {
            lgamma(k + r) - lgamma(k + 1.0) - lgamma(r) + r * log_p + k * log_1mp
        }))
    }

    fn nle(&self, data: &ArrayView1<f64>) -> Option<f64> {
        let n = data.len() as f64;
        if n == 0.0 { return Some(0.0); }

        let sum_y: f64 = data.iter().sum();

        let alpha_post = self.alpha_lambda + sum_y;
        let beta_post = self.beta_lambda + n;

        let log_evidence = lgamma(alpha_post) - lgamma(self.alpha_lambda)
            + self.alpha_lambda * self.beta_lambda.ln()
            - alpha_post * beta_post.ln()
            - data.iter().map(|&k| lgamma(k + 1.0)).sum::<f64>();

        Some(-log_evidence)
    }

    fn nll(&self, data: &ArrayView1<f64>, use_posterior_predictive: bool) -> f64 {
        let n = data.len() as f64;
        if n == 0.0 { return 0.0; }

        let sum_y: f64 = data.iter().sum();

        let alpha_post = self.alpha_lambda + sum_y;
        let beta_post = self.beta_lambda + n;

        if use_posterior_predictive {
            let r = alpha_post;
            let p = beta_post / (1.0 + beta_post);
            let log_p = p.ln();
            let log_1mp = (1.0 - p).ln();

            let mut nll = 0.0;
            for &k in data {
                let log_pmf = lgamma(k + r) - lgamma(k + 1.0) - lgamma(r) + r * log_p + k * log_1mp;
                nll -= log_pmf;
            }
            nll
        } else {
            let lambda = alpha_post / beta_post;
            let log_lambda = lambda.max(1e-10).ln();

            let mut nll = 0.0;
            for &k in data {
                let log_pmf = k * log_lambda - lambda - lgamma(k + 1.0);
                nll -= log_pmf;
            }
            nll
        }
    }

    fn nll_train_test(&self, train: &ArrayView1<f64>, test: &ArrayView1<f64>, use_posterior_predictive: bool) -> f64 {
        let n_train = train.len() as f64;
        let sum_y_train: f64 = train.iter().sum();

        let alpha_post = self.alpha_lambda + sum_y_train;
        let beta_post = self.beta_lambda + n_train;

        if use_posterior_predictive {
            let r = alpha_post;
            let p = beta_post / (1.0 + beta_post);
            let log_p = p.ln();
            let log_1mp = (1.0 - p).ln();

            let mut nll = 0.0;
            for &k in test {
                let log_pmf = lgamma(k + r) - lgamma(k + 1.0) - lgamma(r) + r * log_p + k * log_1mp;
                nll -= log_pmf;
            }
            nll
        } else {
            let lambda = alpha_post / beta_post;
            let log_lambda = lambda.max(1e-10).ln();

            let mut nll = 0.0;
            for &k in test {
                let log_pmf = k * log_lambda - lambda - lgamma(k + 1.0);
                nll -= log_pmf;
            }
            nll
        }
    }

    fn nll_suff_stats(&self, stats: &SufficientStats, use_posterior_predictive: bool) -> Option<f64> {
        let n = stats.n;
        if n == 0.0 { return Some(0.0); }

        let sum_y = stats.sum;

        let alpha_post = self.alpha_lambda + sum_y;
        let beta_post = self.beta_lambda + n;

        if use_posterior_predictive {
            return None;
        } else {
            let lambda = (alpha_post / beta_post).max(1e-10);
            let log_lambda = lambda.ln();
            let nll = n * lambda - sum_y * log_lambda;
            Some(nll)
        }
    }

    fn nle_suff_stats(&self, stats: &SufficientStats) -> Option<f64> {
        let n = stats.n;
        if n == 0.0 { return Some(0.0); }

        let sum_y = stats.sum;

        let alpha_post = self.alpha_lambda + sum_y;
        let beta_post = self.beta_lambda + n;

        let log_evidence = lgamma(alpha_post) - lgamma(self.alpha_lambda)
            + self.alpha_lambda * self.beta_lambda.ln()
            - alpha_post * beta_post.ln();

        Some(-log_evidence)
    }
}
